### DISCUSSION TITLE
Transparent map colors when using Graphics2DFramebuffer

### DISCUSSION THREAD

**Discussion by gabber235** [Minestom/Minestom#692] [Help] [answered]
I want to be able to create a banner board where the background is transparent. Just like seen here:
![image](https://user-images.githubusercontent.com/16163838/154684924-3923213d-e680-4365-8f97-2737a5d44094.png)
https://youtu.be/BIisBmoBQcA

I have tried to recreate the `Graphics2DFramebuffer` but give the buffered image the type of `BufferedImage.TYPE_INT_ARGB`. Which did not help. 
I also tried to set the background colour to `background = Color(255, 255, 255, 0)` which should make it transparent. However, this just results in the specified colour ignoring the alpha value. 

Though with both options enabled, if I save the file to a png, it yields a transparent image. So the Graphics2D is working correctly. It is somewhere on the Minestom side that does not translate it well. 
Is this possible using a `Graphics2DFramebuffer`? If not, is there any other way to accomplish the same effect?

**Comment by jglrxavpok** [comment, marked as answer]
Currently I believe using a background color of R=0, G=0, B=0, A=0 would work. The alpha value is ignored, but the NONE color is set to pure black for maps in Minecraft, so the closest color to your background color would be NONE.

PR https://github.com/Minestom/Minestom/pull/693 may help with that, but will create breaking changes.

**Reply by gabber235** [reply]
Sadly, setting the background colour to R=0, G=0, B=0, A=0 does not work but makes it black. Hopefully, the pr does solve my issue. I will have a look at it tomorrow.

**Reply by gabber235** [reply]
With the PR I managed to get a working version. Hopefully, it will be merged so I can use it natively.

**Comment by jglrxavpok** [comment]
By the way, it should be possible to achieve the same effect if you are willing to use https://github.com/Minestom/LWJGLMapExtension & OpenGL (or Vulkan if you're crazy). 
There is an option to render directly using map colors using `GLFWCapableFramebuffer#useMapColors()`.
But rendering with OpenGL is nowhere as simple as Graphics2D if you've never used it.

**Comment by Tynopia** [comment]
For every people who will use "Graphics2DFramebuffer" and transparent background

```
    @Override
    public byte[] toMapColors() {
        // TODO: update subparts only
        for (int x = 0; x < 128; x++) {
            for (int z = 0; z < 128; z++) {
                int color = get(x, z);

                if (color == 0) {
                    colors[Framebuffer.index(x, z)] = 0x00000000;
                } else {
                    colors[Framebuffer.index(x, z)] = MapColors.closestColor(get(x, z)).getIndex();
                }
            }
        }
        return colors;
    }
```
